### ISSUE TITLE
Code documentation

### ISSUE THREAD

**Issue by tyteen4a03** [serilog/serilog-filters-expressions#25]
I'm trying to adapt this library for non-Serilog purposes, and as a C# newbie I found the code very hard to follow.

It would be nice to have more documentation in the code so I can understand what exactly the code is doing.

**Comment by nblumhardt** [comment]
Hi! Yes, this one's going to be quite a tough nut to crack. We'd like to make the code more approachable, but it's a very big job, so not something we'll be likely to get to in the near future. The techniques in here rest on a whole bunch of background information that would also need to be either documented or called out.

I've recently written a detailed walkthrough of a similar parser in: https://github.com/datalust/superpower/blob/dev/sample/JsonParser/Program.cs - this should definitely help with the parsing side of things.

When it comes to the tree transformations and execution, it might be possible to step through in a debugger to see how things fit together. If you give it a go, and have questions, please feel free to drop us a line here.

Cheers,
Nick

**Comment by tyteen4a03** [comment]
I'm having more problems with the Compiler than the Parser, but thanks for the link. :)

**Comment by nblumhardt** [comment]
:-) :+1:

Will close this here, since I don't think there's much we can do to help right now, but good luck with it!
